core/tracker/tracker.go

```diff
@@ -125,29 +125,73 @@ func (t *Tracker) Run(ctx context.Context) error {
 	}
 }
 
-// analyseDutyFailed detects if a duty failed. It returns true if the duty didn't complete the sigagg component.
-// If it failed, it also returns the component where it failed and a human friendly error message explaining why.
-func analyseDutyFailed(duty core.Duty, es []event) (bool, component, string) {
-	events := make([]event, len(es))
-	copy(events, es)
+// analyseDutyFailed detects if the given duty failed. It returns false if the duty didn't fail, i.e., the duty didn't get stuck and completes the sigAgg component.
+// It returns true if the duty failed. It also returns the component where the duty got stuck and a human friendly error message explaining why.
+func analyseDutyFailed(duty core.Duty, es map[core.Duty][]event) (bool, component, string) {
+	// dutyFailedComponent returns true if the duty failed. It also returns the component where the duty got stuck.  If the duty didn't get stuck, it
+	// returns the sigAgg component. It assumes that all the input events are for a single duty.
+	dutyFailedComponent := func(es []event) (bool, component) {
+		events := make([]event, len(es))
+		copy(events, es)
+
+		// Sort in reverse order (see order above).
+		sort.Slice(events, func(i, j int) bool {
+			return events[i].component > events[j].component
+		})
+
+		if len(events) == 0 {
+			return false, sentinel
+		}
+
+		c := events[0].component
+		if c == sigAgg {
+			// Duty completed successfully
+			return false, sigAgg
+		}
+
+		return true, c
+	}
 
-	// Sort in reverse order (see order above).
-	sort.Slice(events, func(i, j int) bool {
-		return events[i].component > events[j].component
-	})
+	var (
+		failed bool
+		comp   component
+		msg    string
+	)
 
-	if len(events) == 0 {
-		return false, sentinel, "No events to analyse"
+	failed, comp = dutyFailedComponent(es[duty])
```

Comment:
you previously had `comp = events[0].component + 1`, you should probably still do that